packages/editor/src/components/block-removal-warnings/index.js

```diff
@@ -0,0 +1,85 @@
+/**
+ * WordPress dependencies
+ */
+
+import { _n } from '@wordpress/i18n';
+import { privateApis as blockEditorPrivateApis } from '@wordpress/block-editor';
+import { useSelect } from '@wordpress/data';
+import { useMemo } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { unlock } from '../../lock-unlock';
+import { store as editorStore } from '../../store';
+
+const { BlockRemovalWarningModal } = unlock( blockEditorPrivateApis );
+
+// Prevent accidental removal of certain blocks, asking the user for confirmation first.
+const TEMPLATE_BLOCKS = [
+	'core/post-content',
+	'core/post-template',
+	'core/query',
+];
+const BLOCK_REMOVAL_RULES = [
+	{
+		// Template blocks.
+		// The warning is only shown when a user manipulates templates or template parts.
+		postTypes: [ 'wp_template', 'wp_template_part' ],
```

Comment:
Not sure if this message needs to display when editing any other post types than the ones specified here? Or maybe it should be universal (happens for any post type).